plugins/outputs/cloudwatchlogs/internal/pusher/target.go

```diff
@@ -0,0 +1,140 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT
+
+package pusher
+
+import (
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/influxdata/telegraf"
+	"golang.org/x/sync/singleflight"
+
+	"github.com/aws/amazon-cloudwatch-agent/sdk/service/cloudwatchlogs"
+)
+
+type Target struct {
+	Group, Stream, Class string
+	Retention            int
+}
+
+func (t Target) String() string {
+	return fmt.Sprintf("%s:%s:%s:%v", t.Group, t.Stream, t.Class, t.Retention)
+}
+
+type TargetManager interface {
+	InitTarget(target Target) error
+	PutRetentionPolicy(target Target)
+}
+
+type targetManager struct {
+	logger  telegraf.Logger
+	service cloudWatchLogsService
+	// cache of initialized targets
+	cache map[Target]struct{}
+	group singleflight.Group
+}
+
+func NewTargetManager(logger telegraf.Logger, service cloudWatchLogsService) TargetManager {
+	return &targetManager{
+		logger:  logger,
+		service: service,
+		cache:   make(map[Target]struct{}),
+	}
+}
+
+// InitTarget initializes a Target if it hasn't been initialized before.
+func (m *targetManager) InitTarget(target Target) error {
+	_, err, _ := m.group.Do(target.String(), func() (any, error) {
```

Comment:
Does this offer much benefits as opposed to a map with a mutex?

Reply:
No that's a good catch. Was missing a mutex.